sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala

```diff
@@ -198,7 +204,38 @@ case class TransformWithStateExec(
       getOutputRow(obj)
     }
     ImplicitGroupingKeyTracker.removeImplicitKey()
-    mappedIterator
+
+    // Wrapper to ensure that the implicit key is set when the methods on the iterator
+    // are called. Inside of processNewData, we use a GroupedIterator, so handleInputRows
+    // is only called once per key. As such, we only have to set the implicit key when
+    // the first call to hasNext is made, and we have to remove it when hasNext returns
+    // false.
+    //
+    // Note: if we ever start to interleave the processing of the iterators we get back
+    // from handleInputRows (i.e. we don't process each iterator all at once), then this
+    // iterator will need to set/unset the implicit key every time hasNext/next is called,
+    // not just at the first and last calls to hasNext.
+    new Iterator[InternalRow] {
```

Comment:
Why not use NextIterator if we do not consider the case of interleaving now? We can remove the implicit key on completion.

Also let's have a private class implementation of Iterator to deduplicate the code; we wouldn't need to tell to see the above method for code comment in later method.